[core] Add and use application name

marechal-p opened the pull request:
Add customizable application name.
Display application name when no workspace is opened, instead of showing
the URL. It was particulary bad on Electron.

Remove native menus when creating a new Electron BrowserWindow, which will be
re-added by the application when ready.

Signed-off-by: marechal-p <paul.marechal@ericsson.com>

vince-fugnitto commented:
sorry @marechal-p, but I think you used your personal email rather than work :(

vince-fugnitto commented:
it lworks and looks great @marechal-p, perhaps we can simply display `Theia` when no workspace is present? I know vscode does something similar, namely:
- `Visual Studio Code` for no workspaces.
- `${Workspace Name} - Visual Studio Code` when a workspace is present.

akosyakov commented:
@kittaakos can you have a look

I feel a bit unsure since it does additional remote calls at runtime to fetch information which actually is available at the build time. Maybe it should be mixed into frontend config to avoid fetching what is already known.

marechal-p commented:
@akosyakov it was my concern, part of why I waited to merge.

Is there a cache of the props already, or should I make one?

kittaakos commented on packages/workspace/src/browser/workspace-service.ts:
```diff
@@ -168,19 +175,24 @@ export class WorkspaceService implements FrontendApplicationContribution {
-    protected updateTitle(): void {
+    protected async formatTitle(title?: string): Promise<string> {
+        const name = await this.applicationName;
+        return title ? `${title} - ${name}` : name;
```
[minor]: `-` -> `—`? At least, VS Code uses the latter long dash.

kittaakos commented:
> I feel a bit unsure since it does additional remote calls at runtime to fetch information which actually is available at the build time. Maybe it should be mixed into frontend config to avoid fetching what is already known.

I am OK with both approaches. The build-time solution is faster, no questions at all. But fetching the application name once when setting the workspace also seems OK.

marechal-p commented:
TBH I would prefer having the applicationName set at build time too, but while I know how to write something inside the generated sources (modifying the generators), I fail to see a clean way of fetching the value afterward, from any Theia component...

How would you go about it? @akosyakov @kittaakos

kittaakos commented:
> How would you go about it? @akosyakov @kittaakos

Let's spare @akosyakov for the review ;)

I can take a look but then, let's fuse this PR with #3459. What do you think, @marechal-p?

marechal-p commented:
I'll fuze them now

kittaakos commented:
I will create my patch on the [`generate-app-name-patch`](https://github.com/theia-ide/theia/tree/generate-app-name-patch) branch which is a branch from your `mp/startup-window-title`.

marechal-p commented:
Do you allow me to simply fuze your changes inside this PR?

kittaakos commented:
Sure. Use as much as you need.

marechal-p commented:
@kittaakos updated everything accordingly.

- I used the build time string like advised.
- I changed the initial setting of the menus to an empty template instead of `null` because it wasn't working on OSX, it now does.
- Also fixed menu instantiation under OSX to handle the case where we can only set the whole application menu globally, so I hooked the menu creation to each Electron BrowserWindow on the `focus` event. Tested on our OSX machine and it seemed to work, if anyone wants to also give it a try.

marechal-p commented on examples/browser/package.json:
```diff
@@ -2,6 +2,13 @@
+        "applicationName": "Theia Browser Example"
+      }
+    }
+  },
```
Why don't we need this? It gives a good example on how to setup the applicationName, and it clearly displays that example applications are indeed examples.

kittaakos replied:
👍 OK. Let's leave it there as an "example".

marechal-p commented:
> > - I changed the initial setting of the menus to an empty template instead of null because it wasn't working on OSX, it now does.
>
> I am a bit puzzled; your change was already approved twice: #3445 (review) and #3445 (review) and you are saying; it works now. I still do not understand what does this fix.

Exactly what I said. Despite being approved already, it wasn't working on OSX when I tested after (I remembered we had a OSX machine to test on). So I covered that case. It now works on Ubuntu/Windows/OSX. Now corretly hides the default electron menus until the application sets it up.

> - Also fixed menu instantiation under OSX to handle the case where we can only set the whole application menu globally, so I hooked the menu creation to each Electron BrowserWindow on the `focus` event. Tested on our OSX machine and it seemed to work, if anyone wants to also give it a try.

Well, again, on Ubuntu/Windows, we can set the menus on a per-window basis. But not on OSX, where the menu bar is unique and global to all windows, and lives at the top of the screen. Problem is, when you have two electron windows opened, and only one menu bar (OSX), in what window should `New Terminal` open a new terminal in? So my fix sets the "global" OSX menus each time you change windows so that it will target the one you will use, because it is focused.

But I will move that menu change to an other PR, no issues :)

marechal-p commented on packages/core/src/electron-browser/menu/electron-menu-contribution.ts:
```diff
@@ -77,11 +77,23 @@ export class ElectronMenuContribution implements FrontendApplicationContribution
             electron.remote.Menu.setApplicationMenu(createdMenuBar);
+
+            // Update the global menu bar based on the focused Electron BrowserWindow:
+            currentWindow.on('focus', () => {
```
Surely

kittaakos commented:
> Problem is, when you have two electron windows opened, and only one menu bar (OSX), in what window should `New Terminal` open a new terminal in? So my fix sets the "global" OSX menus each time you change windows so that it will target the one you will use, because it is focused.

Thank you for taking the time and explaining it to me again. It makes sense.

marechal-p commented:
@kittaakos it should be better now :)

marechal-p commented:
Ok, I didn't see the failing tests, still addressing that :(

marechal-p commented:
@kittaakos I know this PR is taking longer than it should to get ready, but there is a last point where I want your opinion:

Currently, the application configuration requires us to define the `package.json` in the following way:
```json
{
  "theia": {
    "frontend": {
      "config": {
        "applicationName": "Theia Browser Example"
      }
    }
  }
}
```

Wouldn't it be better to have the application name as a "global", along with the target?
```json
{
  "theia": {
    "applicationName": "Theia Browser Example"
  }
}
```

It sounds rather specific to have it in the frontend config member. On the other hand, only the frontend could care about that, because it is only in the frontend logic that we would try to display this name. The backend would hardly need such a string, or I'm just failing to find a use case for it.

Other than that I'm happy with how the changes are right now, feel free check if everything is ready to merge :)

kittaakos commented:
> Wouldn't it be better to have the application name as a "global", along with the target?

I would leave as is.

> Other than that I'm happy with how the changes are right now, feel free check if everything is ready to merge :)

It looks great! I'll try it out once more on Windows and merge the changes.
